File: python/crypt/rsa.py
#!/usr/bin/env python3
# -*- coding: utf_8 -*-


def exp_pow(x, n, mod):
    res = 1
    while n:
        n, r = divmod(n, 2)
        if r:
            res = (res * x) % mod
        x = x * x % mod
    return res


def factorise(n):
    div = []
    for d in range(2, int(n ** .5)+1):
        while n % d == 0:
            n = n // d
            div.append(d)
    if n != 1:
        div.append(n)
    return div


def pretty_decompose(n):
    div = factorise(n)
    print(n, "=", ".".join([str(i) for i in div]))


def is_prime(n):
    for t in [2, 3, 5, 7, 11, 13, 17]:
        if exp_pow(t, n-1, n) != 1:
            return False
    return len(factorise(n)) == 1

if __name__ == "__main__":
    primes = []
    for n in range(10000001, 20000000, 2):
        if is_prime(n):
            primes.append(n)
    print(len(primes))

File: python/crypt/rsa.cpp
#include <iostream>
#include <vector>
#include <array>
#include <cstdint>
#include <cmath>

using ulll = uintmax_t;

ulll exp_pow(ulll x, ulll n, ulll mod) {
    ulll answer{ 1 }, remainder;
    while(n) {
        remainder = n % 2;
        n /= 2;
        if(remainder)
            answer = (answer * x) % mod;
        x = ((x % mod) * (x % mod)) % mod;
    }
    return answer;
}

std::vector<ulll> factorise(ulll n) {
    std::vector<ulll> div;
    const auto max = static_cast<ulll>(std::sqrt(n)) + 1;
    for(ulll d{2}; d < max; ++d) {
        while(n % d == 0) {
            n /= d;
            div.push_back(d);
        }
    }
    if(n != 1)
        div.push_back(n);
    return div;
}

bool is_prime(ulll n) {
    static constexpr std::array<ulll, 7> tests{
        {2, 3, 5, 7, 11, 13, 17}
    };
    for(auto t : tests)
        if(exp_pow(t, n-1, n) != 1)
            return false;
    return factorise(n).size() == 1;
}

int main() {
    constexpr ulll start = 10000001;
    constexpr ulll end = 20000000;

    std::vector<ulll> primes;
    primes.reserve((end-start)/4);
    for(auto n = start; n < end; n += 2)
        if(is_prime(n))
            primes.push_back(n);
    std::cout << primes.size() << std::endl;
}
